src/main/java/io/logz/jmx2graphite/Jmx2Graphite.java

```diff
@@ -28,10 +28,15 @@
     private final MBeanClient client;
 
     public Jmx2Graphite(Jmx2GraphiteConfiguration conf) {
+        this(conf, false);
+    }
+
+    public Jmx2Graphite(Jmx2GraphiteConfiguration conf, boolean runAsDaemon) {
```

Comment:
Why not have `runAsDaemon` as config option in Jmx2GraphiteConfiguration?

Reply:
I will remove the daemon thread changes, thanks. I think it makes sense to run as a daemon thread when running as a java agent, as otherwise it will prevent the jvm from exiting normally once all the other threads have finished. You can see this by running a simple HelloWorld app that prints out `"hello world"` and then exits - without the daemon changes it will not terminate until you kill it with ctrl-c.

```
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello world!");
    }
}
```

```
$ java -cp /tmp/jmx2graphite/target/jmx2graphite-0-SNAPSHOT-javaagent.jar HelloWorld
Hello world!
```

```
$ java -cp /tmp/jmx2graphite/target/jmx2graphite-0-SNAPSHOT-javaagent.jar -javaagent:/tmp/jmx2graphite/target/jmx2graphite-0-SNAPSHOT-javaagent.jar=GRAPHITE_HOSTNAME=localhost,SERVICE_NAME=hello HelloWorld
22/03/02 09:31:42 INFO jmx2graphite.Jmx2GraphiteJavaAgent: Loading with agentArgument: GRAPHITE_HOSTNAME=localhost,SERVICE_NAME=hello
22/03/02 09:31:42 INFO jmx2graphite.Jmx2Graphite: Running with Mbean client
22/03/02 09:31:42 INFO jmx2graphite.Jmx2GraphiteJavaAgent: Initiated new java agent based Jmx2Graphite instance
22/03/02 09:31:42 INFO jmx2graphite.Jmx2Graphite: Graphite: host = localhost, port = 2004
22/03/02 09:31:42 INFO jmx2graphite.GraphiteClient: Graphite metrics prefix: hello.xxx.
22/03/02 09:31:42 INFO jmx2graphite.GraphiteClient: Graphite Client: using writeTimeoutMs of 21000 [ms]. Establishing connection...
Hello world!
22/03/02 09:31:42 INFO jmx2graphite.MetricsPipeline: Found 28 metric beans. Time = 12ms, for 2022-03-02T09:31:30.000-0500
22/03/02 09:31:42 INFO jmx2graphite.MetricsPipeline: Filtered out 0 metrics out of 28 after white/blacklisting
22/03/02 09:31:43 INFO jmx2graphite.MetricsPipeline: metrics fetched. Time: 244 ms; Metrics: 118
22/03/02 09:31:43 WARN jmx2graphite.MetricsPipeline: Failed writing to Graphite: Failed connecting to Graphite. Error = class java.net.ConnectException: Connection refused (Connection refused)
```
it then hangs until you hit ctrl-c.

Reply:
I think we might need to hook into the shutdown hook and shut it down by default, to have a chance to do proper shutdown

Reply:
yeah, I think the shutdown hook is already registered and should still be invoked.

Reply:
So, if you enableHangupSupport you still have to press ctrl-c?

Reply:
i think behavior on shutdown is not related directly to daemon threads. the shutdown hook will execute when the jvm terminates, regardless of how it terminates (with the caveat that shutdown hooks are not guaranteed). Setting the daemon thread flag just means that the jmx agent will not cause the jvm to stay up when it would otherwise exit. In other words, if the original app required ctrl-c to stop, then it will still require ctrl-c to stop. but if the original app just runs for a while and then exits, then enabling the jmx agent as a daemon thread will preserve that behavior.

Reply:
Since hangup support is suppose to guarantee the stop of those threads, in practice we many no need daemon thread flag, but there is a case where they thread might get stuck on I/O then for that purpose?

Reply:
i think hangup support only applies when the jvm exits, but with the non-daemon threads, the jvm will never exit unless stopped with ctrl-c or similar. It may interrupt the sending of metrics, but that doesn't seem like it would matter much as the jvm is stopping anyway.

Reply:
Hangup support --> shutdown is called -> Executor Service shut called -> All non-daemon threads terminated --> JVM can exit